File: apps/web/src/lib/ai-policy.ts
import type { AIPolicyTier, AIAssistMode } from "@proofpath/types";

export const AI_POLICY_LABELS: Record<AIPolicyTier, string> = {
  locked: "Locked — No AI",
  grammar: "Grammar & Spell Only",
  outline: "Outline Assist",
  chat: "Chat Panel — Query Only",
  inline: "Inline Suggestions",
  full: "Full — Logged",
};

export const AI_POLICY_DESCRIPTIONS: Record<AIPolicyTier, string> = {
  locked: "No AI assistance. Paste from external sources is blocked.",
  grammar: "Grammar and spelling corrections only.",
  outline: "AI may suggest structural outlines, not prose.",
  chat: "Ask questions; responses cannot be inserted into the document.",
  inline: "Copilot-style completions, fully logged.",
  full: "Unrestricted AI use with full interaction logging.",
};

export const AI_POLICY_OPTIONS: AIPolicyTier[] = [
  "locked",
  "grammar",
  "outline",
  "chat",
  "inline",
  "full",
];

export function canUseAIChat(tier: AIPolicyTier): boolean {
  return ["chat", "inline", "full", "grammar", "outline"].includes(tier);
}

export function canUseInline(tier: AIPolicyTier): boolean {
  return ["inline", "full"].includes(tier);
}

export function canUseGrammar(tier: AIPolicyTier): boolean {
  return tier !== "locked";
}

export function canUseOutline(tier: AIPolicyTier): boolean {
  return ["outline", "chat", "inline", "full"].includes(tier);
}

export function isPasteBlocked(tier: AIPolicyTier): boolean {
  return tier === "locked";
}

/** Map policy tier to default AI panel mode */
export function defaultAIMode(tier: AIPolicyTier): AIAssistMode {
  if (tier === "grammar") return "grammar";
  if (tier === "outline") return "outline";
  if (tier === "chat") return "chat";
  return "inline";
}

export function allowedModes(tier: AIPolicyTier): AIAssistMode[] {
  switch (tier) {
    case "locked":
      return [];
    case "grammar":
      return ["grammar"];
    case "outline":
      return ["outline", "chat"];
    case "chat":
      return ["chat"];
    case "inline":
      return ["inline", "chat"];
    case "full":
      return ["grammar", "outline", "chat", "inline"];
    default:
      return [];
  }
}

File: apps/api/app/services/ai_policy.py
"""Server-side AI policy enforcement."""

from fastapi import HTTPException

VALID_POLICIES = frozenset({"locked", "grammar", "outline", "chat", "inline", "full"})

MODE_TO_POLICY: dict[str, frozenset[str]] = {
    "grammar": frozenset({"grammar", "outline", "chat", "inline", "full"}),
    "outline": frozenset({"outline", "chat", "inline", "full"}),
    "chat": frozenset({"chat", "inline", "full"}),
    "inline": frozenset({"inline", "full"}),
}


def validate_policy(tier: str) -> str:
    if tier not in VALID_POLICIES:
        raise HTTPException(400, f"Invalid ai_policy. Must be one of: {', '.join(sorted(VALID_POLICIES))}")
    return tier


def assert_mode_allowed(policy: str, mode: str) -> None:
    if policy == "locked":
        raise HTTPException(403, "AI assistance is disabled for this assignment")
    allowed = MODE_TO_POLICY.get(mode)
    if not allowed or policy not in allowed:
        raise HTTPException(403, f"Mode '{mode}' is not permitted under policy '{policy}'")


def can_insert_text(policy: str) -> bool:
    return policy in ("inline", "full")


def build_system_prompt(
    policy: str,
    assignment_title: str,
    instructions: str | None,
) -> str:
    base = (
        f"You are ProofPath's writing assistant for the assignment: {assignment_title}.\n"
        "Never claim to write the student's work for them. Be calibrated and educational.\n"
    )
    if instructions:
        base += f"\nAssignment instructions:\n{instructions}\n"

    policy_rules = {
        "grammar": (
            "POLICY: Grammar & Spell Only. Only suggest corrections to grammar, spelling, and punctuation. "
            "Do not generate new sentences, paragraphs, or structural content."
        ),
        "outline": (
            "POLICY: Outline Assist. Only provide structural outlines, headings, and bullet points. "
            "Never write full prose paragraphs or complete sentences for the student."
        ),
        "chat": (
            "POLICY: Query Only. Answer questions and explain concepts. "
            "Do not provide text meant to be pasted into the assignment."
        ),
        "inline": (
            "POLICY: Inline Suggestions. Provide short completion suggestions (1-2 sentences max). "
            "Student may insert your suggestions; all interactions are logged."
        ),
        "full": (
            "POLICY: Full Logged. You may help freely but encourage original student work. "
            "All interactions are logged for the authenticity report."
        ),
    }
    return base + "\n" + policy_rules.get(policy, policy_rules["full"])
